src/surb/mod.rs

```diff
@@ -0,0 +1,235 @@
+use crate::constants::{DESTINATION_ADDRESS_LENGTH, PAYLOAD_SIZE, SECURITY_PARAMETER};
+use crate::header::delays::Delay;
+use crate::header::keys::PayloadKey;
+use crate::header::SphinxError;
+use crate::payload::Payload;
+use crate::route::{Destination, Node, NodeAddressBytes};
+use crate::{crypto, header, SphinxPacket};
+use curve25519_dalek::scalar::Scalar;
+
+#[derive(Clone)]
+pub struct SURB {
+    /// A Single Use Reply Block (SURB) must have a pre-aggregated Sphinx header,
+    /// the address of the first hop in the route of the SURB, and the key material
+    /// used to layer encrypt the payload.
+    pub SURBHeader: header::SphinxHeader,
+    pub first_hop_address: NodeAddressBytes,
+    pub payload_keys: Vec<PayloadKey>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum SURBError {
+    IncorrectSURBRoute,
+}
+
+impl SURB {
+    pub fn new(
+        surb_initial_secret: Scalar,
+        surb_route: &[Node],
+        surb_delays: &[Delay],
+        surb_destination: &Destination,
+    ) -> Result<Self, SURBError> {
+        /// Precomputes the header of the Sphinx packet which will be used as SURB
+        /// and encapsulates it into struct together with the address of the first hop in the route of the SURB, and the key material
+        /// which should be used to layer encrypt the payload.
+        assert_eq!(surb_route.len(), surb_delays.len());
```

Comment:
Do an explicit check for `if surb_route.len() == surb_delays.len()`, rather than an assertion, because if assertion fails, the code will panic rather than return an Err